### ISSUE TITLE
Using RuleBasedStateMachine for tests that have a final state

### ISSUE THREAD

**Issue by bheisler** [HypothesisWorks/hypothesis#2403]
Hi folks. I've been trying to use Hypothesis's RuleBasedStateMachine tests to do something unusual and I'm hoping you can help me out. I have a complex, event-driven piece of code that essentially implements a state machine - as events happen, the "state machine" transitions to new states and sends additional messages, which can then trigger additional events and so on.

For the most part, using Hypothesis for this works great (and the resulting Hypothesis test is easier to read and more powerful than a similar example-based test would have been), but there is one thing that I haven't been able to do.

The state machine that I'm testing has a single final state that all paths ultimately terminate at. I want to use Hypothesis to assert that the implementation does always reach that state within a reasonable number of steps (plus additional assertions about the state of the simulated world when it does). The way I've structured this test, each possible event is a rule (using preconditions to ensure that Hypothesis only generates events that could actually happen at that point in time). The problem is I can't figure out how to write the assertion that it always reaches the final state - if Hypothesis doesn't perform enough steps then it won't reach the final state. On the other hand, if Hypothesis performs too many steps then when it does reach the final state and there are no more valid events available, then it raises an error ("No progress can be made from state") and considers the test failed. In this construction, that's a success - there are no further events to process and the system has correctly halted.

The way I've worked around this for now is to set a flag when the system reaches its final state and have an empty rule that can only run when that flag is set. This way Hypothesis can run that rule over and over until the number of steps is exhausted. However, I can't find any way to assert that the system has reached that final state - if the number of steps is too low, it might not ever test that part of the code. I've changed the settings to ratchet up the number of steps, but I don't know if it's high enough.

As another workaround, I tried putting that assertion in `teardown`, but that's clearly not the intended way to do it so I'm not surprised that it didn't work.

Caveat: I'm running Hypothesis 4.23.4 on Python 2.7 (I know...). I know Python 2.7 is no longer supported. Even if you add some new feature to support this use case, I probably won't be able to use it for a while. I still think it would be a useful feature, and I hope that I'll be able to use it once my employer updates to Python 3. I've looked through the documentation for the current version and I didn't see anything that fit my use case, but if I've just missed it then please let me know.

Additional details: I don't know if my simplified explanation of my use case above is comprehensible, so I'll try to unpack it a bit more. This code is essentially a dispatcher - it receives a fixed set of units of work up front and passes each one to a worker for processing. It then takes those results as they become available (it uses futures to register callbacks to be triggered when the processing is complete) and optionally passes them to a second worker for a different sort of processing, then back to the first, and so on until all of the work is completed. In production, the two workers are separate processes which might be running on different machines; for the sake of unit testing I've replaced those with fake implementations of each worker that store the messages in a list. I then use Hypothesis to pop a random message out of the list and handle it. I want to ensure that the system always reaches a state where all of the units of work have been completed or failed, and none of them have been forgotten. In this case, that's represented by the different processing queues being empty as well as a message sent by the dispatcher when all of the units of work have finished.

**Comment by Zac-HD** [comment]
I think you might actually be better off with the `st.data()` strategy than stateful testing - in particular, the challenges around how many steps to run make it a poor fit for RuleBasedStateMachine which can run as few steps as it likes.  Sketching out the alternative,

```python
@given(data=st.data(), events=an_input_queue())
def test_foo(...):
    for _ in range(max_steps):
        ... # Choose and run a step here
        if done():  break
    else:
        assert False, "too many steps"
    assert reached_end_state() and queues_are_empty()
```
    
My condolences on the old versions - you might at least be able to use the final 4.x release?  I appreciate that you're interested in helping support a use-case that you won't immediately get to use, too - thanks!

**Comment by bheisler** [comment]
That's a good idea, thanks. I tried it; there are a few limitations:
- It means that I have to build my own version of `@rule` and `@precondition`. I ended up combining them into a single `@event` for my particular case just for simplicity.
- The output when a test fails is not quite as nice, since it shows only a list of values drawn from the `data()` object rather than a list of the rule functions that were called. I was mostly able to clear this up with some additional work by adding `__repr__()` functions to eg. print the name of the event function that was called instead of the usual function object repr string.

In this way, I was able to test what I wanted. It works really well, actually; I introduced a bug in the code I was testing to verify that the test works and Hypothesis found it and generated a minimal reproduction right away.

I'll leave this open in case you want to use it to track any changes related to this.
